File: backend/app/engine/generate.py
"""
Core pattern + fabric calculations (Python port of frontend/lib/pattern-engine/generate.ts).
Pure functions — no I/O, easy to unit test.
"""

from dataclasses import dataclass, field
from typing import List, Optional

from app.engine.rules import get_rule

AVG_ADULT_HEIGHT_INCH = 64.0  # ~163cm baseline for fabric-length scaling


@dataclass
class MeasurementInput:
    bust: Optional[float] = None
    waist: Optional[float] = None
    hip: Optional[float] = None
    shoulder: Optional[float] = None
    arm_round: Optional[float] = None
    sleeve_length: Optional[float] = None
    neck: Optional[float] = None
    dress_length: Optional[float] = None
    height: Optional[float] = None
    unit: str = "inch"


@dataclass
class PatternPiece:
    name: str
    width: float
    height: float
    notes: str


@dataclass
class PatternResult:
    unit: str
    seam_allowance: float
    dart_width: float
    dart_length: float
    pieces: List[PatternPiece]
    warnings: List[str] = field(default_factory=list)


@dataclass
class FabricResult:
    main_fabric: float
    lining: float
    lace: float
    elastic: float
    interfacing: float
    wastage_percent: int
    total_with_wastage: float


def _to_inches(value: float, unit: str) -> float:
    return value / 2.54 if unit == "cm" else value


def generate_pattern(dress_name: str, m: MeasurementInput) -> PatternResult:
    rule = get_rule(dress_name)
    unit = m.unit or "inch"
    warnings: List[str] = []

    bust = m.bust or 0
    waist = m.waist or 0
    shoulder = m.shoulder or 0
    arm_round = m.arm_round or 0
    sleeve_length = m.sleeve_length or 0
    neck = m.neck or 0
    dress_length = m.dress_length or 0

    if not bust:
        warnings.append("Bust measurement missing — front/back width defaulted to 0. Add it for an accurate draft.")
    if not dress_length:
        warnings.append("Dress length missing — using 0. Add it to size the body pieces.")

    seam_allowance = rule.seam_allowance
    quarter_bust = bust / 4
    ease = rule.ease_factor / 4

    front_width = quarter_bust + ease + seam_allowance * 2
    back_width = quarter_bust - 0.25 + seam_allowance * 2

    armhole_depth = (shoulder / 2 + 2.25) if shoulder else (bust / 6 + 2.5)
    body_length = dress_length + rule.hem_allowance + seam_allowance

    dart_width = max(bust / 4 - waist / 4, 0.5)
    dart_length = armhole_depth * 0.65

    neck_width = (neck / 5) if neck else 2.5
    neck_depth = (neck / 5 + 0.5) if neck else 3.0

    sleeve_width = (arm_round / 2 + 1.5 + seam_allowance * 2) if arm_round else quarter_bust
    sleeve_full_length = sleeve_length + seam_allowance + rule.hem_allowance / 2

    pieces = [
        PatternPiece(
            "Front Body", round(front_width, 1), round(body_length, 1),
            f"Neck opening {round(neck_width, 1)}×{round(neck_depth, 1)} {unit}, cut on fold",
        ),
        PatternPiece(
            "Back Body", round(back_width, 1), round(body_length, 1),
            f"Dart: {round(dart_width, 1)} {unit} wide × {round(dart_length, 1)} {unit} long, cut on fold",
        ),
        PatternPiece(
            "Sleeve", round(sleeve_width, 1), round(sleeve_full_length, 1),
            f"Cut 2 (mirror pair), ease into armhole depth {round(armhole_depth, 1)} {unit}",
        ),
    ]

    return PatternResult(unit, seam_allowance, round(dart_width, 1), round(dart_length, 1), pieces, warnings)


def calculate_fabric(dress_name: str, m: MeasurementInput, base_fabric_meters: float) -> FabricResult:
    rule = get_rule(dress_name)
    unit = m.unit or "inch"

    height_inch = _to_inches(m.height, unit) if m.height else AVG_ADULT_HEIGHT_INCH
    waist_inch = _to_inches(m.waist, unit) if m.waist else 0

    height_scale = min(max(height_inch / AVG_ADULT_HEIGHT_INCH, 0.75), 1.35)
    main_fabric = base_fabric_meters * rule.fabric_scale * height_scale

    lining = main_fabric * rule.lining_factor if rule.needs_lining else 0
    lace = rule.lace_factor if rule.needs_lace else 0
    elastic = ((waist_inch or (163 / 2.54 / 2.5)) * 0.0254 * rule.elastic_factor) if rule.needs_elastic else 0
    interfacing = rule.interfacing_meters if rule.needs_interfacing else 0

    subtotal = main_fabric + lining
    total_with_wastage = subtotal * (1 + rule.wastage)

    return FabricResult(
        main_fabric=round(main_fabric, 2),
        lining=round(lining, 2),
        lace=round(lace, 2),
        elastic=round(elastic, 2),
        interfacing=round(interfacing, 2),
        wastage_percent=round(rule.wastage * 100),
        total_with_wastage=round(total_with_wastage, 2),
    )

File: lib/pattern-engine/generate.ts
import { getRule, type DressRule } from "./rules";

export type MeasurementInput = {
  bust?: number | null;
  waist?: number | null;
  hip?: number | null;
  shoulder?: number | null;
  armRound?: number | null;
  sleeveLength?: number | null;
  neck?: number | null;
  dressLength?: number | null;
  height?: number | null;
  unit?: "inch" | "cm";
};

export type PatternPiece = {
  name: string;
  width: number;
  height: number;
  notes: string;
};

export type PatternResult = {
  unit: "inch" | "cm";
  seamAllowance: number;
  dartWidth: number;
  dartLength: number;
  pieces: PatternPiece[];
  warnings: string[];
};

export type FabricResult = {
  unit: "meter";
  mainFabric: number;
  lining: number;
  lace: number;
  elastic: number;
  interfacing: number;
  wastagePercent: number;
  totalWithWastage: number;
};

const AVG_ADULT_HEIGHT_INCH = 64; // ~163cm, used as the baseline for scaling base fabric
const AVG_ADULT_HEIGHT_CM = 163;

function toInches(value: number, unit: "inch" | "cm") {
  return unit === "cm" ? value / 2.54 : value;
}

/**
 * Computes simplified front/back/sleeve pattern dimensions from body
 * measurements using standard-ease drafting formulas. Values are in the
 * same unit the measurements were recorded in.
 */
export function generatePattern(dressName: string, m: MeasurementInput): PatternResult {
  const rule = getRule(dressName);
  const unit = m.unit ?? "inch";
  const warnings: string[] = [];

  const bust = m.bust ?? 0;
  const waist = m.waist ?? 0;
  const shoulder = m.shoulder ?? 0;
  const armRound = m.armRound ?? 0;
  const sleeveLength = m.sleeveLength ?? 0;
  const neck = m.neck ?? 0;
  const dressLength = m.dressLength ?? 0;

  if (!bust) warnings.push("Bust measurement missing — front/back width defaulted to 0. Add it for an accurate draft.");
  if (!dressLength) warnings.push("Dress length missing — using 0. Add it to size the body pieces.");

  const seamAllowance = rule.seamAllowance;
  const quarterBust = bust / 4;
  const ease = rule.easeFactor / 4;

  // Front is slightly wider than back to account for bust shaping
  const frontWidth = quarterBust + ease + seamAllowance * 2;
  const backWidth = quarterBust - 0.25 + seamAllowance * 2;

  const armholeDepth = shoulder ? shoulder / 2 + 2.25 : bust / 6 + 2.5;
  const bodyLength = dressLength + rule.hemAllowance + seamAllowance;

  const dartWidth = Math.max(bust / 4 - waist / 4, 0.5);
  const dartLength = armholeDepth * 0.65;

  const neckWidth = neck ? neck / 5 : 2.5;
  const neckDepth = neck ? neck / 5 + 0.5 : 3;

  const sleeveWidth = armRound ? armRound / 2 + 1.5 + seamAllowance * 2 : quarterBust;
  const sleeveFullLength = sleeveLength + seamAllowance + rule.hemAllowance / 2;

  const pieces: PatternPiece[] = [
    {
      name: "Front Body",
      width: round1(frontWidth),
      height: round1(bodyLength),
      notes: `Neck opening ${round1(neckWidth)}×${round1(neckDepth)} ${unit}, cut on fold`,
    },
    {
      name: "Back Body",
      width: round1(backWidth),
      height: round1(bodyLength),
      notes: `Dart: ${round1(dartWidth)} ${unit} wide × ${round1(dartLength)} ${unit} long, cut on fold`,
    },
    {
      name: "Sleeve",
      width: round1(sleeveWidth),
      height: round1(sleeveFullLength),
      notes: `Cut 2 (mirror pair), ease into armhole depth ${round1(armholeDepth)} ${unit}`,
    },
  ];

  return { unit, seamAllowance, dartWidth: round1(dartWidth), dartLength: round1(dartLength), pieces, warnings };
}

/**
 * Computes total fabric requirement (main fabric, lining, lace, elastic,
 * interfacing) in meters, including standard cutting wastage.
 */
export function calculateFabric(
  dressName: string,
  m: MeasurementInput,
  baseFabricMeters: number
): FabricResult {
  const rule = getRule(dressName);
  const unit = m.unit ?? "inch";

  const heightInch = m.height ? toInches(m.height, unit) : AVG_ADULT_HEIGHT_INCH;
  const waistInch = m.waist ? toInches(m.waist, unit) : 0;

  const heightScale = clamp(heightInch / AVG_ADULT_HEIGHT_INCH, 0.75, 1.35);
  const mainFabric = baseFabricMeters * rule.fabricScale * heightScale;

  const lining = rule.needsLining ? mainFabric * rule.liningFactor : 0;
  const lace = rule.needsLace ? rule.laceFactor : 0;
  const elastic = rule.needsElastic
    ? (waistInch || AVG_ADULT_HEIGHT_CM / 2.54 / 2.5) * 0.0254 * rule.elasticFactor
    : 0;
  const interfacing = rule.needsInterfacing ? rule.interfacingMeters : 0;

  const subtotal = mainFabric + lining;
  const totalWithWastage = subtotal * (1 + rule.wastage);

  return {
    unit: "meter",
    mainFabric: round2(mainFabric),
    lining: round2(lining),
    lace: round2(lace),
    elastic: round2(elastic),
    interfacing: round2(interfacing),
    wastagePercent: Math.round(rule.wastage * 100),
    totalWithWastage: round2(totalWithWastage),
  };
}

/** Static, per-dress-type cutting guidance — folding, sequence, and fabric-saving tips. */
export function getCuttingGuide(dressName: string): {
  foldingInstructions: string;
  steps: string[];
  fabricSavingTips: string[];
} {
  const rule = getRule(dressName);
  const foldingInstructions =
    rule.category === "kids"
      ? "Fold fabric in half lengthwise, selvedge to selvedge. Kids' pieces are small — nest them inside the offcuts of the body pieces."
      : "Fold fabric in half lengthwise (selvedge to selvedge) for symmetrical pieces (front/back on the fold); open a single layer for the sleeve pieces to cut 2 in mirror.";

  const steps = [
    "Press the fabric flat before cutting to remove fold lines that distort measurements.",
    "Lay the Front Body piece on the fold first — it's the widest piece and anchors the layout.",
    "Place the Back Body piece alongside, sharing the same fold line where possible.",
    "Cut the Sleeve pieces from the remaining single-layer fabric, mirrored (cut 2).",
    rule.needsLining ? "Cut lining pieces from the lining fabric using the same Front/Back pattern, minus seam allowance at the hem." : "Finish raw edges with overlock or a folded hem — no lining required for this style.",
    rule.needsLace ? "Reserve lace for the neckline and sleeve hems — measure and cut after the body pieces are stitched, to match the actual finished edge." : "Skip lace unless requested as a design add-on.",
    "Mark all notches (shoulder, side seam, sleeve head) with tailor's chalk before lifting pieces off the fabric.",
  ];

  const fabricSavingTips = [
    "Cut the sleeves from the fabric left over after the body pieces, rather than reserving a separate length.",
    "Alternate the direction of pattern pieces (head-to-toe) on printed/striped fabric only if the print is non-directional.",
    `Standard wastage for ${dressName} is budgeted at ${Math.round(rule.wastage * 100)}% — buy that much extra, no more.`,
    "Keep offcuts larger than 6 inches — they're enough for facings, plackets, or a matching mask/scrunchie for referral upsells.",
  ];

  return { foldingInstructions, steps, fabricSavingTips };
}

function round1(n: number) {
  return Math.round(n * 10) / 10;
}
function round2(n: number) {
  return Math.round(n * 100) / 100;
}
function clamp(n: number, min: number, max: number) {
  return Math.min(Math.max(n, min), max);
}

export type { DressRule };
